File: tools/dirtytooth.c
/**
 *
 *  Copyright (C) 2017  Raul Casanova Marques
 *
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include <assert.h>
#include <getopt.h>

#include <bluetooth/bluetooth.h>

#include "../src/pbap-types.h"
#include "../src/pbap.h"

void write_to_file(const void* filename, const void* buf, size_t buflen);

static struct option long_options[] = {
		{"interface", required_argument, 0, 'i'},
		{"address", required_argument, 0, 'a'},
		{"channel", required_argument, 0, 'c'},
		{"help", no_argument, 0, 'h'},
		{0, 0, 0, 0}
};

int main(int argc, char* argv[])
{
	struct pbap_t pbap;

	unsigned char* buf;
	size_t buflen;

	uint16_t dev_id;
	uint8_t channel;
	char* addr;

	int opt;

	/* default values */
	dev_id = 0;
	channel = 0xFF;
	addr = NULL;

	while ((opt = getopt_long(argc, argv, "i:a:c:h", long_options, NULL)) != - 1)
	{
		switch (opt)
		{
			case 'i':
			{
				/* check dev, ex: hci0 */
				assert(strlen(optarg) == 4);
				dev_id = (uint16_t) atoi(optarg + 3);
				break;
			}
			case 'a':
			{
				/* check bluetooth address */
				assert(bachk(optarg) == 0);
				addr = strdup(optarg);

				break;
			}
			case 'c':
			{
				channel = (uint8_t) atoi(optarg);
				break;
			}
			case 'h':
			{
				fprintf(stderr, "Usage: %s [--interface=<hciX>] --address=<00:00:00:00:00:00> --channel=<XX>\n", argv[0]);

				break;
			}
			default:
			{
				break;
			}
		}
	}

	/* check required arguments */
	if (channel == 0xFF || addr == NULL)
	{
		fprintf(stderr, "Usage: %s [--interface=<hciX>] --address=<00:00:00:00:00:00> --channel=<XX>\n", argv[0]);
		return - 1;
	}

	/* start pbap client */
	pbap = pbap_client(dev_id, addr, channel);
	if (pbap.pbap_status == PBAP_INIT_OK)
	{
		/* pull phone book */
		buflen = pbap_get(pbap, TELECOM_PB, sizeof(TELECOM_PB), XBT_PHONEBOOK, sizeof(XBT_PHONEBOOK), (void*) &buf);
		write_to_file("pb.vcf", buf, buflen);

		/* pull call history */
		buflen = pbap_get(pbap, TELECOM_CCH, sizeof(TELECOM_CCH), XBT_PHONEBOOK, sizeof(XBT_PHONEBOOK), (void*) &buf);
		write_to_file("cch.vcf", buf, buflen);

		pbap_cleanup(pbap);
	}

	return 0;
}

void write_to_file(const void* filename, const void* buf, size_t buflen)
{
	FILE* f;
	size_t i;

	f = fopen(filename, "wt");
	if (f != NULL)
	{
		/* write buffer to the file */
		for (i = 0; i < buflen; i ++)
		{
			fprintf(f, "%c", ((unsigned char*) buf)[i]);
		}
		/* close the file */
		fclose(f);
	}
	else
	{
		perror("fopen");
	}
}

File: test/dirtytooth.py
#!/usr/bin/python

#
#  Copyright (C) 2017  Raul Casanova Marques
#
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import datetime
import sys
import time

import bluetooth
from nOBEX import client, headers, responses

device_address = '00:00:00:00:00:00'


def write_file(filename, data):
    with open(filename, 'w') as f:
        f.write(data)


def get_file(c, src_path, filename, book=True):
    if book:
        mimetype = b'x-bt/phonebook'
    else:
        mimetype = b'x-bt/vcard'

    hdrs, data = c.get(src_path, header_list=[headers.Type(mimetype)])
    write_file(filename, data)
    print('Done!')


def main():
    svc = bluetooth.find_service(address=device_address, uuid='1130')
    if len(svc) < 1:
        sys.stderr.write('Failed to find PBAP service.\n')
        sys.exit(1)

    channel = svc[0]['port']
    c = client.Client(device_address, channel)

    # 796135f0-f0c5-11d8-0966-0800200c9a66
    uuid = b'\x79\x61\x35\xf0\xf0\xc5\x11\xd8\x09\x66\x08\x00\x20\x0c\x9a\x66'
    r = c.connect(header_list=[headers.Target(uuid)])

    if not isinstance(r, responses.ConnectSuccess):
        sys.stderr.write('Failed to connect to phone.\n')
        sys.exit(1)

    date = datetime.datetime.fromtimestamp(time.time()).strftime('%Y%m%d%H%M%S')

    # Phone book
    get_file(c, 'telecom/pb.vcf', '%s_%s-UTC_phonebook' % (device_address, date))

    # Call history
    get_file(c, 'telecom/cch.vcf', '%s_%s-UTC_history' % (device_address, date))

    c.disconnect()


if __name__ == '__main__':
    main()
